SetFlags.cmake

```diff
@@ -233,87 +81,81 @@ macro(enable_profile)
 	endif()
 endmacro()
 
-macro(set_exe_flags)
-	# Remove disabling the maximum warning level:
-	# clang does not like a command line that reads -Wall -Wextra -w -Wall -Wextra and does not output any warnings
-	# We do not do that for MSVC since MSVC produces an awful lot of warnings for its own STL headers;
-	# the important warnings are turned on using #pragma in Globals.h
-	if (NOT MSVC)
-		string(REPLACE "-w" "" CMAKE_CXX_FLAGS_RELEASE  "${CMAKE_CXX_FLAGS_RELEASE}")
-		string(REPLACE "-w" "" CMAKE_C_FLAGS_RELEASE    "${CMAKE_C_FLAGS_RELEASE}")
-		string(REPLACE "-w" "" CMAKE_CXX_FLAGS_DEBUG    "${CMAKE_CXX_FLAGS_DEBUG}")
-		string(REPLACE "-w" "" CMAKE_C_FLAGS_DEBUG      "${CMAKE_C_FLAGS_DEBUG}")
-		string(REPLACE "-w" "" CMAKE_CXX_FLAGS_COVERAGE "${CMAKE_CXX_FLAGS_COVERAGE}")
-		string(REPLACE "-w" "" CMAKE_C_FLAGS_COVERAGE   "${CMAKE_C_FLAGS_COVERAGE}")
-		add_flags_cxx("-Wall -Wextra -Wno-unused-parameter -Wno-missing-noreturn")
+# Add coverage processing, if requested:
+if (NOT MSVC)
 
-		# we support non-IEEE 754 fpus so can make no guarentees about error
-		add_flags_cxx("-ffast-math")
+	if (CMAKE_BUILD_TYPE STREQUAL "COVERAGE")
+		message("Including CodeCoverage")
+		set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/lib/cmake-coverage/")
+		include(CodeCoverage)
+	endif()
+endif()
+
+# TODO: is this needed? NDEBUG is standard. Also, why are we using _DEBUG?
+# Add the preprocessor macros used for distinguishing between debug and release builds (CMake does this automatically for MSVC):
+if (NOT MSVC)
+	set(CMAKE_CXX_FLAGS_DEBUG    "${CMAKE_CXX_FLAGS_DEBUG}    -D_DEBUG")
+	set(CMAKE_C_FLAGS_DEBUG      "${CMAKE_C_FLAGS_DEBUG}      -D_DEBUG")
+	set(CMAKE_CXX_FLAGS_COVERAGE "${CMAKE_CXX_FLAGS_COVERAGE} -D_DEBUG")
+	set(CMAKE_C_FLAGS_COVERAGE   "${CMAKE_C_FLAGS_COVERAGE}   -D_DEBUG")
+	set(CMAKE_CXX_FLAGS_RELEASE  "${CMAKE_CXX_FLAGS_RELEASE}  -DNDEBUG")
+	set(CMAKE_C_FLAGS_RELEASE    "${CMAKE_C_FLAGS_RELEASE}    -DNDEBUG")
+endif()
+
+if(MSVC)
+	# Make build use multiple threads under MSVC:
+	add_compile_options(/MP)
+
+	# Make build use Unicode:
+	add_compile_definitions(UNICODE _UNICODE)
+
+	# TODO: level 4, warnings as errors
+else()
+	target_compile_options(
+		${CMAKE_PROJECT_NAME} PRIVATE
```

Comment:
Instead of assuming the target name, it would be better if you changed the signature to
```cmake
macro(set_exe_flags target)
```

And likewise for the other macros I think. Better to be explicit about what targets we are effecting.